src/serac/physics/utilities/variational_form/finite_element.hpp

```diff
@@ -0,0 +1,181 @@
+   */
+  constexpr operator int() { return d; }
+};
+
+/**
+ * @brief Returns the dimension of an element geometry
+ * @param[in] g The @p Geometry to retrieve the dimension of
+ */
+constexpr int dimension_of(Geometry g)
+{
+  if (g == Geometry::Segment) {
+    return 1;
+  }
+
+  if (g == Geometry::Triangle || g == Geometry::Quadrilateral) {
+    return 2;
+  }
+
+  if (g == Geometry::Tetrahedron || g == Geometry::Hexahedron) {
+    return 3;
+  }
+
+  return -1;
+}
+
+/**
+ * @brief H1 elements of order @p p
+ * @tparam p The order of the elements
+ * @tparam c The vector dimension
+ */
+template <int p, int c = 1>
+struct H1 {
+  static constexpr int order      = p;
+  static constexpr int components = c;
+};
+
+/**
+ * @brief H(curl) elements of order @p p
+ * @tparam p The order of the elements
+ * @tparam c The vector dimension
+ */
+template <int p, int c = 1>
+struct Hcurl {
+  static constexpr int order      = p;
+  static constexpr int components = c;
+};
+
+/**
+ * @brief Discontinuous elements of order @p p
+ * @tparam p The order of the elements
+ * @tparam c The vector dimension
+ */
+template <int p, int c = 1>
+struct L2 {
+  static constexpr int order      = p;
+  static constexpr int components = c;
+};
+
+/**
+ * @brief Element conformity
+ */
+enum class Family
+{
+  H1,
+  HCURL,
+  HDIV,
+  L2
+};
+
+/**
+ * @brief FIXME: This doesn't appear to be used anywhere
+ */
+enum class Evaluation
+{
+  Interpolate,
+  Divergence,
+  Gradient,
+  Curl
+};
+
+/**
+ * @brief Template prototype for finite element implementations
+ * @tparam g The geometry of the element
+ * @tparam family The continuity of the element
+ * the implementations of the different finite element specializations
+ * are in .inl files in the detail/ directory. 
+ * 
+ * In each of these files, the finite_element specialization
+ * should implement the following concept:
+ * 
+ * struct finite_element< some_geometry, some_space > > {
+ *   static constexpr Geometry geometry = ...;
+ *   static constexpr Family family     = ...;
+ *   static constexpr int  components   = ...;
+ *   static constexpr int  dim          = ...;
+ *   static constexpr int  ndof         = ...;
+ * 
+ *   static constexpr auto shape_functions(tensor<double, dim> xi) { ... }
+ * 
+ *   static constexpr auto shape_function_derivatives(tensor<double, dim> xi) { ... }
+ * };
+ *  
+ */
+template <Geometry g, typename family>
+struct finite_element;
+
+/**
+ * @brief Type trait for identifying finite element types
+ */
+template <typename T>
+struct is_finite_element {
+  static constexpr bool value = false;
+};
+/// @overload
+template <Geometry g, int p, int c>
+struct is_finite_element<finite_element<g, H1<p, c> > > {
+  static constexpr bool value = true;
+};
+/// @overload
+template <Geometry g, int p>
+struct is_finite_element<finite_element<g, Hcurl<p> > > {
+  static constexpr bool value = true;
+};
+
+#include "detail/segment_h1.inl"
+//#include "detail/segment_hcurl.inl"
```

Comment:
A todo statement is needed to explain why these are commented out.

Reply:
these are implemented and uncommented in the weak_form_L2 branch, maybe I should merge that in to this one